Add a `Flaky failures` tutorial docs page (#4482)
HypothesisWorks/hypothesis: hypothesis-python/docs/tutorial/flaky.rst

Flaky failures
==============

Have you ever had a test fail, and then you re-run it, only for the test to magically pass? That is a *flaky test*. A flaky test is one which might behave differently when called again. You can think of it as a test which is not deterministic.

Any test can be flaky, but because Hypothesis runs your test many times, Hypothesis tests are particularly likely to uncover flaky behavior.

> Note that Hypothesis does not require tests to be fully deterministic. Only the sequence of calls to Hypothesis APIs like ``draw`` from |st.composite| and the outcome of the test (pass or fail) need to be deterministic. This means you can use randomness, threads, or nondeterminism in your test, as long as it doesn't impact anything Hypothesis can see. We jokingly call this requirement "morally deterministic".

Reviewer: I'd delete the morally-deterministic quip, it's a good one especially when talking to FP people about pure functions, but a little too niche for a tutorial 😢

Reviewer 2: I'm ambivalent, I like it and think it's clear even to non-FP folks, but I might be too far away from a novice at this point to write a good tutorial. I removed it

* Hypothesis relies on deterministic behavior for the database to work.
* Flakiness makes debugging failures substantially harder if the failing input reported by Hypothesis only flakily reproduces.
* Flakiness makes effectively exploration of the test's behavior space by Hypothesis difficult or impossible.

Common sources of flakiness
---------------------------

Here is a quick and non-exhaustive enumeration of some reasons you might encounter flakiness:

* Decisions based on global state.
* Explicit dependencies between inputs.
* Test depends on filesystem or database state which isn't reset between inputs.
* Un-managed sources of randomness. This includes standard PRNGs (see also |register_random|), but also thread scheduling, network timing, etc.

.. note::

    If your tests depend on global state, consider replacing that state with |st.shared|. This is a common way to refactor your test to bring conceptually-global state under the control and visibility of Hypothesis.

Types of flakiness
------------------

When Hypothesis detects that a test is flaky, it will raise one of two |Flaky| exceptions.

Flaky failure
~~~~~~~~~~~~~

> The most common form of flakiness is that Hypothesis finds a failure, but then replaying it does not reproduce that failure. For example, here is a contrived test which only fails the first time it is called:

Reviewer: ```suggestion
The most common form of flakiness is that Hypothesis finds a failure, but then replaying that input does not reproduce that failure. For example, here is a contrived test which only fails the first time it is called:
```

As a result, running ``test_fails_flakily()`` will raise |FlakyFailure|. |FlakyFailure| is an ``ExceptionGroup``, which contains the origin failure as a sub-exception:

.. code-block:: none

  + Exception Group Traceback (most recent call last):
  | hypothesis.errors.FlakyFailure: Hypothesis test_fails_flakily(n=0) produces unreliable results: Falsified on the first call but did not on a subsequent one (1 sub-exception)
  | Falsifying example: test_fails_flakily(
  |     n=0,
  | )
  | Failed to reproduce exception. Expected:
  | Traceback (most recent call last):
  |   File "/Users/tybug/Desktop/sandbox2.py", line 13, in test_fails_flakily
  |     assert False
  |            ^^^^^
  | AssertionError
  +-+---------------- 1 ----------------
    | ...
    | Traceback (most recent call last):
    |   File "/Users/tybug/Desktop/sandbox2.py", line 13, in test_fails_flakily
    |     assert False
    |            ^^^^^
    | AssertionError
    +------------------------------------

The solution to seeing |FlakyFailure| is to refactor the test to not depend on external state. In this case, the external state is the variable ``called``.

> Flaky strategy
> ~~~~~~~~~~~~~~

Reviewer: ```suggestion
Flaky strategy definition
~~~~~~~~~~~~~~~~~~~~~~~~~
```

> A more fundamental form of flakiness is if the strategy's data generation itself is flaky (or not deterministic). Since Hypothesis relies on this to replay and shrink failing inputs, it requires that data generation is not flaky.
> 
> One easy way for this to occur is if a strategy depends on external state. For example, this strategy generates a unique integer each iteration:
> 
> .. code-block:: python
> 
>     seen = set()
> 
>     @st.composite
>     def unique_ints(draw):
>         while (n := draw(st.integers())) in seen:
>             pass
>         seen.add(n)
>         return n
> 
>     @given(unique_ints())
>     def test_ints(n): ...
> 
> By using ``seen``, this test is relying on outside state! On the first iteration where |st.integers| generates ``0``, ``unique_ints`` draws only one integer. But on the next iteration where |st.integers| generates ``0``, ``unique_ints`` draws two integers. This means data generation is not deterministic.
> 
> As a result, running ``test_ints()`` will raise |FlakyStrategyDefinition|. The solution is to refactor the strategy to not depend on external state. One way to do this is using |st.shared|.

Reviewer: This is pretty subtle, and I'm wary enough of the words "flaky" and "unique" being misunderstood that I rewrote a bunch of it.  Also added a worked example of using shared.

```suggestion
Each strategy must 'do the same thing' (again, as seen by Hypothesis) if we replay a previously-seen input.  Failing to do so is a more subtle but equally serious form of flakiness, which leaves us unable to shrink to a minimal failing example, or even reliably report the failure in future runs.

One easy way for this to occur is if a strategy depends on external state. For example, this strategy filters out previously-generated integers, including those seen in any previous test case:

.. code-block:: python

    seen = set()

    @st.composite
    def unique_ints(draw):
        while (n := draw(st.integers())) in seen:
            pass
        seen.add(n)
        return n

    @given(unique_ints(), unique_ints())
    def test_ints(x, y): ...

By using ``seen``, this test is relying on outside state! In the first test case where |st.integers| generates ``0``, ``unique_ints`` draws only one integer. But if in the next test case |st.integers| generates ``0``, ``unique_ints`` has to draw two integers because ``0`` is already in ``seen``. This means data generation is not deterministic.

As a result, running ``test_ints()`` will raise |FlakyStrategyDefinition|. The solution is to refactor the strategy to not depend on external state. One way to do this is using |st.shared|:

.. code-block:: python

    @st.composite
    def unique_ints(draw):
        seen_this_time = draw(st.shared(st.builds(set), key="seen_ints"))
        while (n := draw(st.integers())) in seen_this_time:
            pass
        seen_this_time.add(n)
        return n
```

Reviewer 2: nice! I reworded a few small things

Flaky failures
==============

Have you ever had a test fail, and then you re-run it, only for the test to magically pass? That is a *flaky test*. A flaky test is one which behaves differently when called again; you can think of it as a test which is not deterministic.

> Any test can be flaky, but because Hypothesis runs your test many times, Hypothesis tests are particularly likely to discover flakiness.

Reviewer: - or passes and then fails!  And 'behaves _or might behave_ differently'.  (not wordsmithing, just concept)
- I think we want to emphasize that there are multiple reasons to avoid flakiness.
  - it's a poor user experience, wastes time, and makes debugging much harder if you can't trust a repro.
  - Hypothesis relies on being able to 'replay' a test for the database to work
  - and also for effective exploration of test behavior!
- importantly "the same behavior" doesn't require _literally everything_ to be the same: just that, if Hypothesis makes the same sequence of choices, then your test will make the same sequence of calls to Hypothesis apis, and raise (or not) the same exception.

A line that might be useful: this means that tests _do not_ need to be "pure functions", only what we joking call 'morally pure'.

Reviewer 2: > or passes and then fails! And 'behaves or might behave differently'. (not wordsmithing, just concept)

idk, I don't particularly want to rigorously define flakiness here. I think the current wording gets the intuitive concept across concisely. I do like `behave` -> `might behave` though

agree everywhere else

            called = True
            assert False

The first time Hypothesis generates an input, this test will fail. But when Hypothesis tries replaying that failure—by generating the same input—the test will succeed. This test is flaky.

As a result, running ``test_fails_flakily()`` will raise |FlakyFailure|. |FlakyFailure| is an ``ExceptionGroup``, which contains the origin failure as a sub-exception:

.. code-block:: none

  + Exception Group Traceback (most recent call last):
  | hypothesis.errors.FlakyFailure: Hypothesis test_fails_flakily(n=0) produces unreliable results: Falsified on the first call but did not on a subsequent one (1 sub-exception)
  | Falsifying example: test_fails_flakily(
  |     n=0,
  | )
  | Failed to reproduce exception. Expected:
  | Traceback (most recent call last):
  |   File "/Users/tybug/Desktop/sandbox2.py", line 13, in test_fails_flakily
  |     assert False
  |            ^^^^^
  | AssertionError
  +-+---------------- 1 ----------------
    | ...
    | Traceback (most recent call last):
    |   File "/Users/tybug/Desktop/sandbox2.py", line 13, in test_fails_flakily
    |     assert False
    |            ^^^^^
    | AssertionError

>     +------------------------------------

Reviewer: (probably for later)  I want to go through the possibilities here, because I suspect that printing the expected traceback is redundant with the newer exceptiongroup approach.

The solution to seeing |FlakyFailure| is to refactor the test to not depend on external state. In this case, the external state is the variable ``called``.

Flaky strategy
~~~~~~~~~~~~~~

A more fundamental form of flakiness is if the strategy's data generation itself is flaky (or not deterministic). Since Hypothesis relies on this to replay and shrink failures, it requires that data generation is not flaky.

> One easy way for this to occur is if a strategy depends on external state. For example, this strategy generates a unique integer each iteration:

Reviewer: Feels helpful if we can enumerate some other ways:

- decisions based on global state / explicit depenencies between test cases
- test depends on e.g. filesystem or database state, which isn't reset between inputs (c.f. Django extra, fixtures healthcheck)
- some source of unmanaged randomness, e.g. thread scheduling or network timing or etc.  (see also register_random)

which I guess is maybe general rather than strategy-level 🤔

Reviewer 2: I put this top-level, I don't love it, open to alternative orderings

.. code-block:: python

    seen = set()

    @st.composite
    def unique_ints(draw):

>         while (n := draw(st.integers())) in seen:
>             seen.add(n)
>         return n

Reviewer: ```suggestion
        while (n := draw(st.integers())) in seen:
            pass
        seen.add(n)
        return n
```

    @given(unique_ints())
    def test_ints(n): ...

By using ``seen``, this test is relying on outside state! On the first iteration where |st.integers| generates ``0``, ``unique_ints`` draws only one integer. But on the next iteration where |st.integers| generates ``0``, ``unique_ints`` draws two integers. This means data generation is not deterministic between generated inputs.

> As a result, running ``test_ints()`` will raise |FlakyStrategyDefinition|. The solution is to refactor the strategy to not depend on external state. (in the case of ``unique_ints``, it can simply be replaced with |st.integers|, as Hypothesis already deduplicates generated inputs across iterations).

Reviewer: Eh, no, we do sometimes produce duplicate values as an explicit flakiness check, in addition to cases where (in more complex strategies) you can produce a particular input from multiple underlying choice sequences.

I'd also take this opportunity to point users who are thinking about global state to `st.shared()` for the "test-case-scoped dict" trick.  Do that instead of `global n`, and the strategy would be fine!

Reviewer 2: yeah I'm not really a fan of this example in general tbh. feels to specific and contrived to be useful

Reviewer 2: I've removed this quote:

> (in the case of ``unique_ints``, it can simply be replaced with |st.integers|, as Hypothesis already deduplicates generated inputs across iterations).

and added a comment about st.shared
